pkg/datapath/l2responder/l2responder.go

```diff
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package l2responder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime/pprof"
+	"time"
+
+	"github.com/cilium/cilium/pkg/datapath/tables"
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+	"github.com/cilium/cilium/pkg/hive/job"
+	"github.com/cilium/cilium/pkg/maps/l2respondermap"
+	"github.com/cilium/cilium/pkg/statedb"
+	"github.com/cilium/cilium/pkg/types"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+var Cell = cell.Module("l2-responder", "L2 Responder Reconciler",
+	cell.Invoke(NewL2ResponderReconciler),
+	cell.Provide(newNeighborNetlink),
+)
+
+type params struct {
+	cell.In
+
+	Lifecycle           hive.Lifecycle
+	Logger              logrus.FieldLogger
+	L2AnnouncementTable statedb.Table[*tables.L2AnnounceEntry]
+	StateDB             statedb.DB
+	L2ResponderMap      l2respondermap.Map
+	NetLink             linkByNamer
+	JobRegistry         job.Registry
+}
+
+type linkByNamer interface {
+	LinkByName(name string) (netlink.Link, error)
+}
+
+func newNeighborNetlink() linkByNamer {
+	return &netlink.Handle{}
+}
+
+type l2ResponderReconciler struct {
+	params params
+}
+
+func NewL2ResponderReconciler(params params) *l2ResponderReconciler {
+	reconciler := l2ResponderReconciler{
+		params: params,
+	}
+
+	group := params.JobRegistry.NewGroup(
+		job.WithLogger(params.Logger),
+		job.WithPprofLabels(pprof.Labels("cell", "l2-responder-reconciler")),
+	)
+	params.Lifecycle.Append(group)
+	group.Add(job.OneShot("l2-responder-reconciler", reconciler.run))
+
+	return &reconciler
+}
+
+func (p *l2ResponderReconciler) run(ctx context.Context) error {
+	log := p.params.Logger
+
+	// This timer triggers full reconciliation once in a while, in case partial reconciliation
+	// got out of sync or the map was changed underneath us.
+	ticker := time.NewTicker(5 * time.Minute)
+
+	// At startup, do an initial full reconciliation
+	maxRev, err := p.fullReconciliation()
+	if err != nil {
+		log.WithError(err).Error("Error(s) while reconciling l2 responder map")
+	}
+
+	for ctx.Err() == nil {
+		maxRev = p.cycle(ctx, maxRev, ticker.C)
```

Comment:
maxRev here, and above never gets used?

Reply:
maxRev is also an argument. cycle is called in a loop feeding the output from the last call to the next. It is used to track revision numbers for the incremental/partial update

Reply:
Ah it's passed into the arguments I see.